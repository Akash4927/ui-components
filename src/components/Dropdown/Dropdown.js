import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { map, find, filter, includes, lowerCase, concat } from 'lodash';

import Input from '../Input';
import Text from '../Text';

const WIDTH = '256px';
const HEIGHT = '36px';

const Item = styled.div`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 24px;
  cursor: pointer;
`;

const Items = styled.div`
  max-height: 400px;
  overflow-y: auto;
`;

const Popover = styled.div`
  position: absolute;
  background-color: ${props => props.theme.colors.white};
  border: 1px solid ${props => props.theme.colors.neutral.lightgray};
  border-radius: ${props => props.theme.borderRadius};
  z-index: 4;
  box-shadow: ${props => props.theme.boxShadow.light};
  margin-top: 4px;
  width: ${WIDTH};
  box-sizing: border-box;
`;

const Overlay = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  z-index: 2;
`;

const ItemWrapper = Item.extend`
  line-height: ${HEIGHT};
  color: ${props =>
    props.selected ? props.theme.colors.accent.blue : props.theme.textColor};
  min-height: ${HEIGHT};
  cursor: ${props => (props.unclickable ? 'not-allowed' : 'pointer')};
  background-color: ${props => (props.softSelected ? '#eee' : 'inherit')};

  &:hover {
    background-color: ${props => (props.unclickable ? 'inherit' : '#eee')};
  }

  &:first-child {
    border-radius: ${props => props.theme.borderRadius} 0 0;
  }

  &:last-child {
    border-bottom: 0;
    border-radius: 0 0 ${props => props.theme.borderRadius};
  }
`;

const SelectedItem = Item.extend`
  border-radius: ${props => props.theme.borderRadius};
  background-color: ${props => props.theme.colors.neutral.white};
  border: 1px solid ${props => props.theme.colors.neutral.gray};
  display: flex;
  position: relative;
  z-index: 3;

  ${Item} {
    padding: 0;
  }

  div:last-child {
    margin: auto 0 auto auto;
  }

  ${Input} {
    padding: 0;
    & i,
    span {
      display: none;
    }

    & > div {
      margin: 0 auto;
    }

    input {
      box-sizing: border-box;
      width: 100%;
      line-height: 24px;
      padding: 0 12px;
    }
  }
`;

const SelectedItemIcon = styled.span`
  padding-left: 1em;
`;

const OtherOptions = styled.div`
  border-top: 1px solid ${props => props.theme.colors.neutral.gray};
`;

const StyledDropdown = component => styled(component)`
  line-height: ${HEIGHT};
  position: relative;
  width: ${WIDTH};
`;

const Label = styled.div`
  line-height: initial;

  ${Text} {
    font-size: 12px;
  }
`;

function filterItems(items, query) {
  if (!query) {
    return items;
  }

  const result = filter(items, i =>
    includes(lowerCase(i.label), lowerCase(query))
  );
  if (result.length === 0) {
    return [{ value: null, label: 'No items found' }];
  }
  return result;
}

function findCurrentItem(items, value) {
  return find(items, i => i.value === value) || (items && items[0]);
}

// For keyboard navigation
function moveSoftSelect({ softSelectedIndex, items }, amt) {
  const next =
    typeof softSelectedIndex === 'number' ? softSelectedIndex + amt : 0;

  // Keep the index within the upper bound of the array
  if (next > items.length - 1) {
    return softSelectedIndex;
  }

  // lower bound of the array
  if (next < 0) {
    return 0;
  }

  return next;
}

/**
 * A selectable drop-down menu.
 * ```javascript
 *const items = [
 *  {
 *    value: 'first-thing',
 *    label: 'First Thing',
 *  },
 *  {
 *    value: 'second-thing',
 *    label: 'Second Thing',
 *  },
 *  {
 *    value: 'third-thing',
 *    label: 'Super long thing, this get truncated',
 *  },
 * ];
 *
 * <Dropdown items={items} />
 * ```
 */

class Dropdown extends React.Component {
  constructor(props, context) {
    super(props, context);
    const { items, query, value } = this.props;

    this.state = {
      isOpen: false,
      query: '',
      softSelectedIndex: null,
      items: filterItems(items, query),
      currentItem: findCurrentItem(items, value)
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleToggle = this.handleToggle.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
    this.handleKeyboardNavigation = this.handleKeyboardNavigation.bind(this);
    this.filterItems = this.filterItems.bind(this);
  }

  componentDidMount() {
    this.filterItems(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.filterItems(nextProps);
  }

  filterItems({ items, searchable, value, otherOptions }) {
    this.setState({
      items:
        searchable && this.state.query
          ? filterItems(items, this.state.query)
          : items,
      currentItem: findCurrentItem(concat(items, otherOptions), value)
    });
  }

  handleChange(ev, value) {
    this.setState({ isOpen: false });
    if (this.props.onChange) {
      this.props.onChange(ev, value);
    }
  }

  handleToggle() {
    this.setState(prevState => ({ isOpen: !prevState.isOpen, query: '' }));
  }

  handleSearch(ev) {
    const query = ev.target.value;
    this.setState({
      query,
      items: filterItems(this.props.items, query),
      softSelectedIndex: null
    });
  }

  handleKeyboardNavigation(ev) {
    switch (ev.key) {
      case 'ArrowDown':
        ev.preventDefault();
        this.setState(prevState => ({
          softSelectedIndex: moveSoftSelect(prevState, 1)
        }));
        break;
      case 'ArrowUp':
        ev.preventDefault();
        this.setState(prevState => ({
          softSelectedIndex: moveSoftSelect(prevState, -1)
        }));
        break;
      case 'Enter':
        ev.preventDefault();
        this.handleChange(
          ev,
          this.state.items[this.state.softSelectedIndex].value
        );
        break;
      case 'Escape':
        this.handleToggle();
        break;
      default:
    }
  }

  render() {
    const { value, className, otherOptions, searchable, label } = this.props;
    const { isOpen, items, currentItem, softSelectedIndex } = this.state;

    return (
      <div className={className}>
        <Label>
          <Text>{label}</Text>
        </Label>
        <SelectedItem onClick={this.handleToggle}>
          {searchable && isOpen ? (
            <Input
              focus
              onChange={this.handleSearch}
              onClick={e => e.stopPropagation()}
              onKeyDown={this.handleKeyboardNavigation}
            />
          ) : (
            <Item title={currentItem.label}>
              {currentItem && currentItem.label}
            </Item>
          )}

          <div>
            <SelectedItemIcon
              className={searchable ? 'fa fa-search' : 'fa fa-caret-down'}
            />
          </div>
        </SelectedItem>
        {isOpen && (
          <div>
            <Popover>
              <Items>
                {map(items, (item, index) => (
                  <ItemWrapper
                    className="dropdown-item"
                    key={item.value}
                    unclickable={!item.value}
                    onClick={ev =>
                      item.value && this.handleChange(ev, item.value)
                    }
                    selected={item.value === value}
                    softSelected={index === softSelectedIndex}
                    title={item.label}
                  >
                    <Text italic={!item.value}>{item.label}</Text>
                  </ItemWrapper>
                ))}
              </Items>
              {otherOptions &&
                otherOptions.length > 0 && (
                  <OtherOptions>
                    {map(otherOptions, o => (
                      <ItemWrapper
                        onClick={ev => this.handleChange(ev, o.value)}
                        key={o.value}
                        title={o.label}
                      >
                        {o.label}
                      </ItemWrapper>
                    ))}
                  </OtherOptions>
                )}
            </Popover>
            <Overlay onClick={this.handleToggle} />
          </div>
        )}
      </div>
    );
  }
}

Dropdown.propTypes = {
  /**
   * Array of items that will be selectable. `value` should be an internal value,
   * `label` is what will be displayed to the user.
   */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string
    })
  ).isRequired,
  /**
   * The value of the currently selected item. This much match a value in the `items` prop.
   * If no value is provided, the first elements's value will be used.
   */
  value: PropTypes.string,
  /**
   * A handler function that will run when a value is selected.
   */
  onChange: PropTypes.func,
  /**
   * A list of other options that will appear at the bottom of the list.
   * Should have the same `value` and `label` fields that the `items` prop has.
   */
  otherOptions: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string
    })
  ),
  /**
   * If searchable, the dropdown will render an input box and filter out items in the dropdown list.
   */
  searchable: PropTypes.bool
};

export default StyledDropdown(Dropdown);
