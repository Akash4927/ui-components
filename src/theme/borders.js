import { kebabCase, forEach } from 'lodash';

export const borderRadius = {
  none: '0',
  soft: '2px',
  circle: '50%',
};

// Collects all theme border radii as SCSS vars.
export function themeBorderRadiiAsScss() {
  const themeBorderRadii = [];

  forEach(borderRadius, (value, name) => {
    themeBorderRadii.push(`$border-radius-${kebabCase(name)}: ${value}`);
  });

  return themeBorderRadii;
}
