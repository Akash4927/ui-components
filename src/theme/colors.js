import { isString, kebabCase, forEach } from 'lodash';

export const colors = {
  // uncategorized colors
  aliceBlue: '#f0f8ff',
  amethystSmoke: '#9292b7',
  athens: '#e2e2ec',
  athensGray: '#eeeef4',
  black: '#1a1a1a',
  blueHaze: '#c0c0d5',
  comet: '#5b5b88',
  cometDark: '#5a5a7b',
  doveGray: '#737373',
  eastBay: '#4a5d87',
  gallery: '#f4f4f4',
  gunpowder: '#3c3c5b',
  kimberly: '#7d7da8',
  kimberlyDark: '#656597',
  logan: '#b1b1cb',
  mischka: '#dfdfea',
  mulledWine: '#46466a',
  mystic: '#dae4ea',
  silver: '#ccc',
  turquoise: '#037aa9',
  white: '#fff',
  whiteLilac: '#f0f0fa',
  whiteSmoke: '#fafafc',
  promQL: {
    /**
     * GHColors theme by Avi Aryan (http://aviaryan.in)
     * Inspired by Github syntax coloring
     */
    comment: '#bbbbbb',
    string: '#e3116c',
    punctuation: '#393a34',
    entity: '#36acaa',
    metricName: '#2aa198',
    attrName: '#00a4db',
    function: '#dc322f',
    deleted: '#9a050f',
    tag: '#00009f',
    // Dropdown colors,
    salmon: '#ff7c7c',
  },
  // use these for new stuff!
  primary: {
    charcoal: '#32324b',
    lavender: '#8383ac',
  },
  accent: {
    blue: '#00d2ff',
    orange: '#ff4b19',
  },
  status: {
    // Use these when colorizing elements that should indicate a state or outcome, ie Alerts.
    success: '#38bd93', // green
    error: '#c7254e', // red
    warning: '#d4ab27', // amber
    info: '#049cd7', // blue
    disabled: '#9a9a9a',
  },
  // Used by PrometheusGraph component
  graphThemes: {
    blue: [
      '#c7e9b4',
      '#7ecdbb',
      '#1eb5eb',
      '#1d91bf',
      '#235fa9',
      '#253393',
      '#084181',
    ],
    purple: [
      '#c1d4e7',
      '#9fbddb',
      '#8d95c6',
      '#8282ab',
      '#89429e',
      '#800f7a',
      '#0b0533',
    ],
    mixed: [
      '#c7e9b4',
      '#c1d4e7',
      '#7ecdbb',
      '#9fbddb',
      '#1eb5eb',
      '#8d95c6',
      '#1d91bf',
      '#8282ab',
      '#235fa9',
      '#89429e',
      '#253393',
      '#800f7a',
      '#084181',
      '#0b0533',
    ],
  },
  // Third-party specific colors - not to be used in the theme!
  thirdParty: {
    // Google single-click login
    cornflowerBlue: '#4285f4',
    azure: '#3769bb',
  },
};

// Flattens and collects all theme colors as SCSS vars
export function themeColorsAsScss() {
  const ignoreKeys = ['graphThemes'];
  const themeColors = [];

  forEach(colors, (value, name) => {
    const colorPrefix = `$color-${kebabCase(name)}`;
    if (ignoreKeys.includes(name)) return;

    if (isString(value)) {
      themeColors.push(`${colorPrefix}: ${value}`);
    } else {
      forEach(value, (innerValue, subname) => {
        if (ignoreKeys.includes(subname)) return;
        themeColors.push(`${colorPrefix}-${kebabCase(subname)}: ${innerValue}`);
      });
    }
  });

  return themeColors;
}
