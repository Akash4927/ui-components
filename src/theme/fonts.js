import { kebabCase, forEach } from 'lodash';

export const fontFamilies = {
  regular: "'proxima-nova', Helvetica, Arial, sans-serif",
  monospace: "'Roboto Mono', monospace",
};

export const fontSizes = {
  huge: '48px',
  extraLarge: '32px',
  large: '22px',
  normal: '16px',
  small: '14px',
  tiny: '12px',
};

// Collects all theme font vars as SCSS vars.
export function themeFontsAsScss() {
  const themeFonts = [];

  forEach(fontFamilies, (value, name) => {
    themeFonts.push(`$font-family-${kebabCase(name)}: ${value}`);
  });

  forEach(fontSizes, (value, name) => {
    themeFonts.push(`$font-size-${kebabCase(name)}: ${value}`);
  });

  return themeFonts;
}
