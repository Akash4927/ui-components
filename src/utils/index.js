import intersection from 'lodash/intersection';
import first from 'lodash/first';
import keys from 'lodash/keys';

// Return the value of a set of props against a lookup object.
// Helps when using "ternary" props, ie: <Text small bold />
// Third argument is the default value to return in the event of an `undefined`;
export const lookupValue = (props, lookup, _default) => {
  const matches = intersection(keys(props), keys(lookup));

  if (matches.length > 1) {
    throw new Error(`You have specified conflicting props: ${JSON.stringify(matches)}`);
  }

  return lookup[first(matches)] || _default;
};
