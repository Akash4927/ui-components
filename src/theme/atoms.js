import { darken } from 'polished';

import { colors } from './colors';

// component-specific
export const atoms = {
  Alert: {
    default: {
      color: colors.black,
      background: colors.gallery,
      borderColor: darken(0.15, colors.gallery),
    },
    info: {
      color: colors.white,
      background: colors.status.info,
      borderColor: darken(0.15, colors.status.info),
    },
    success: {
      color: colors.white,
      background: colors.status.success,
      borderColor: darken(0.15, colors.status.success),
    },
    warning: {
      color: colors.white,
      background: colors.status.warning,
      borderColor: darken(0.15, colors.status.warning),
    },
    error: {
      color: colors.white,
      background: colors.status.error,
      borderColor: darken(0.15, colors.status.error),
    },
  },
  Button: {
    default: {
      color: colors.black,
      background: colors.white,
      hoverColor: colors.purple800,
      hoverBackground: darken(0.05, colors.white),
    },
    primary: {
      color: colors.white,
      background: colors.turquoise,
      hoverColor: colors.white,
      hoverBackground: darken(0.05, colors.turquoise),
    },
    danger: {
      color: colors.white,
      background: colors.status.error,
      hoverColor: colors.white,
      hoverBackground: darken(0.05, colors.status.error),
    },
    disabled: {
      color: colors.doveGray,
      background: colors.gallery,
      hoverColor: colors.doveGray,
      hoverBackground: colors.gallery,
    },
  },
};
