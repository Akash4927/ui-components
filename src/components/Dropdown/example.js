import React from 'react';

import { Example } from '../../utils/example';
import Dropdown from '.';

const items = [
  {
    value: 'first-thing',
    label: 'First Thing',
  },
  {
    value: 'second-thing',
    label: 'Second Thing',
  },
  {
    value: 'third-thing',
    label: 'Super long thing, this should get truncated',
  },
];

export default class DropdownExample extends React.Component {
  constructor() {
    super();
    this.state = {
      selected: items[0].value,
    };
  }

  render() {
    const onChange = (ev, value) => {
      this.setState(() => ({ selected: value }));
    };

    return (
      <div>
        <Example>
          <Dropdown items={items} value={this.state.selected} onChange={onChange} />
        </Example>
      </div>
    );
  }
}
