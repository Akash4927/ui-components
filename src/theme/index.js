import { colors, themeColorsAsScss } from './colors';
import { fontSizes, fontFamilies, themeFontsAsScss } from './fonts';
import { borderRadius, themeBorderRadiiAsScss } from './borders';
import { layers, themeLayersAsScss } from './layers';
import { spacing, themeSpacingsAsScss } from './spacings';
import { overlayIconSize, textColor, themeMiscVarsAsScss } from './misc';
import { boxShadow } from './box-shadows';
import { atoms } from './atoms';

export function themeVarsAsScss() {
  const themeVariables = []
    .concat(themeColorsAsScss())
    .concat(themeLayersAsScss())
    .concat(themeFontsAsScss())
    .concat(themeBorderRadiiAsScss())
    .concat(themeSpacingsAsScss())
    .concat(themeMiscVarsAsScss());
  return `${themeVariables.join('; ')};`;
}

const theme = {
  // Component-specific
  atoms,

  // Box shadows
  boxShadow,

  // Borders
  borderRadius,

  // Colors
  colors,

  // Fonts
  fontFamilies,
  fontSizes,

  // z-index layers
  layers,

  // Spacings
  spacing,

  // Misc
  overlayIconSize,
  textColor,
};

export default theme;
