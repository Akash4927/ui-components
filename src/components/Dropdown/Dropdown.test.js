import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { withTheme } from '../../utils/theme';

import Dropdown from './Dropdown';

describe('<Dropdown />', () => {
  const items = [
    {
      value: 'first-thing',
      label: 'First Thing',
    },
    {
      value: 'second-thing',
      label: 'Second Thing',
    },
    {
      value: 'third-thing',
      label: 'Super long thing, this should get truncated',
    },
  ];

  describe('snapshots', () => {
    it('renders default', () => {
      const tree = renderer
        .create(withTheme(<Dropdown items={items} />))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('is extendable', () => {
    // Make sure a styled-component gets exported
    expect(typeof Dropdown.extend).toEqual('function');
    expect(Dropdown.displayName).toEqual('Dropdown');
  });
});
