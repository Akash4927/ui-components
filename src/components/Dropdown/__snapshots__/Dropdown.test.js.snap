// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> snapshots renders a placeholder 1`] = `
.c3 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 12px;
  cursor: pointer;
}

.c1 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 12px;
  cursor: pointer;
  border-radius: 4px;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(0,0%,45%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .c2 {
  padding: 0;
}

.c1 div:last-child {
  margin-left: auto;
}

.c4 {
  padding-left: 1em;
  float: right;
  line-height: 36px !important;
}

.c0 {
  height: 36px;
  line-height: 36px;
  position: relative;
  padding: 8px;
  width: 256px;
}

<div
  className="c0"
  title="Select an item"
>
  <div
    className="c1"
    onClick={[Function]}
  >
    <div
      className="c2 c3"
    >
      Select an item
    </div>
    <div>
      <span
        className="fa fa-caret-down c4"
      />
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> snapshots renders default 1`] = `
.c3 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 12px;
  cursor: pointer;
}

.c1 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 12px;
  cursor: pointer;
  border-radius: 4px;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(0,0%,45%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .c2 {
  padding: 0;
}

.c1 div:last-child {
  margin-left: auto;
}

.c4 {
  padding-left: 1em;
  float: right;
  line-height: 36px !important;
}

.c0 {
  height: 36px;
  line-height: 36px;
  position: relative;
  padding: 8px;
  width: 256px;
}

<div
  className="c0"
  title="First Thing"
>
  <div
    className="c1"
    onClick={[Function]}
  >
    <div
      className="c2 c3"
    >
      First Thing
    </div>
    <div>
      <span
        className="fa fa-caret-down c4"
      />
    </div>
  </div>
</div>
`;
